OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs

```diff
@@ -0,0 +1,149 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.IO;
+using System.Linq;
+using OpenRA.FileFormats;
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.SpriteLoaders
+{
+	public class PngSheetLoader : ISpriteLoader
+	{
+		class PngSheetFrame : ISpriteFrame
+		{
+			public Size Size { get; set; }
+			public Size FrameSize { get; set; }
+			public float2 Offset { get; set; }
+			public byte[] Data { get; set; }
+			public bool DisableExportPadding { get { return false; } }
+		}
+
+		public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
+		{
+			if (!Png.Verify(s))
+			{
+				frames = null;
+				return false;
+			}
+
+			var png = new Png(s);
+			Rectangle[] frameRegions;
+			float2[] frameOffsets;
+
+			// Prefer manual defined regions over auto sliced regions.
+			if (png.Meta.Any(meta => meta.Key.StartsWith("Frame[")))
+				RegionsFromFrames(png, out frameRegions, out frameOffsets);
+			else
+				RegionsFromSlices(png, out frameRegions, out frameOffsets);
+
+			frames = new ISpriteFrame[frameRegions.Length];
+
+			for (var i = 0; i < frames.Length; i++)
+			{
+				frames[i] = new PngSheetFrame()
+				{
+					Size = new Size(frameRegions[i].Width, frameRegions[i].Height),
+					FrameSize = new Size(frameRegions[i].Width, frameRegions[i].Height),
+					Offset = frameOffsets[i],
+					Data = new byte[frameRegions[i].Width * frameRegions[i].Height]
+				};
+
+				var frameStart = frameRegions[i].X + frameRegions[i].Y * png.Width;
+
+				for (var y = 0; y < frames[i].Size.Height; y++)
+					Array.Copy(png.Data, frameStart + y * png.Width, frames[i].Data, y * frames[i].Size.Width, frames[i].Size.Width);
+			}
+
+			return true;
+		}
+
+		void RegionsFromFrames(Png png, out Rectangle[] frameRegions, out float2[] frameOffsets)
+		{
+			var regions = new List<Rectangle>();
+			var offsets = new List<float2>();
+
+			string frame;
+			for (var i = 0; png.Meta.TryGetValue("Frame[" + i + "]", out frame); i++)
+			{
+				// Format: x,y,width,height,offsetX,offsetY
+				var coords = frame.Split(',');
+				regions.Add(new Rectangle(int.Parse(coords[0]), int.Parse(coords[1]), int.Parse(coords[2]), int.Parse(coords[3])));
+				offsets.Add(new float2(int.Parse(coords[4]), int.Parse(coords[5])));
```

Comment:
This is the only tricky case, can we use a different delimiter (e.g. `;` between the rect and offsets, then split on that and give the result to FieldLoader for these?